Add blog post to reflect state of PodReadyToStartContainersCondition (#43932)
kubernetes/website: content/en/blog/_posts/2023-11-14-PodReadyToStartContainersCondition-in-beta.md

Reviewer: /hold

OK to unhold once the date (in the path, and also in the front matter at the top of the Markdown) is 2023-12-14.

Author: updated to 2023-12-14

---
layout: blog
title: "Kubernetes 1.29: PodReadyToStartContainersCondition in beta"
date: 2023-11-13
slug: pod-ready-to-start-containers-condition-now-in-beta
---

**Author**: Kevin Hannon(Red Hat)  zefeng chen

> With Kubernetes 1.29 the PodReadyToStartContainersCondition functionality entered beta and is available by default.

Reviewer 2: ```suggestion
With Kubernetes 1.29 the PodReadyToStartContainers functionality entered beta and is available by default.
```

Reviewer: My version:
```suggestion
With the recent release of Kubernetes 1.29, the `PodReadyToStartContainers`
[condition](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) is available by default,
and the kubelet manages the value for that condition throughout a Pod's lifecycle.
```

Author: updated. did a little combine

> A new kubelet-managed pod condition in the status field of a pod. The kubelet will use the PodHasNetwork condition to accurately surface the initialization state of a pod,

Reviewer 2: ```suggestion
A new kubelet-managed pod condition in the status field of a pod. The kubelet will use the PodReadyToStartContainers condition to accurately surface the initialization state of a pod,
```

Reviewer: Style nit: condition names should be inside backticks. This helps people tell them apart from API kinds (which we write without backticks).

Author: updated Kevin's comment. 

> Style nit: condition names should be inside backticks. This helps people tell them apart from API kinds (which we write without backticks).

can you list me an example? I'm not quite getting this. sorry for my poor English

---
layout: blog
title: "Kubernetes 1.29: PodReadyToStartContainersCondition in beta"
date: 2023-11-13
slug: pod-ready-to-start-containers-condition-now-in-beta
---

**Author**: zefeng chen, Kevin Hannon(Red Hat)

With recent release of Kubernetes 1.29 the `PodReadyToStartContainers`[condition](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) is available by default.

> The kubelet manages the value for that condition throughout a Pod's lifecycle, in the status field of a pod. The kubelet will use the PodReadyToStartContainers condition to accurately surface the initialization state of a pod,

Reviewer 3: ```suggestion
The kubelet manages the value for that condition throughout a Pod's lifecycle,
in the status field of a pod. The kubelet will use the PodReadyToStartContainers
condition to accurately surface the initialization state of a pod,
```

Author: updated.

Cluster administrators currently do not have a clear and easily accessible way to view the status of successfully scheduled pods 
in terms of the completion of pod sandbox creation and initialization. The current Initialized condition in pods tracks the execution of init containers. 
However, it has limitations in accurately reflecting the completion of sandbox creation and readiness to start containers for all pods in a cluster. 
This distinction is particularly important in multi-tenant clusters where tenants own the pod specifications, including the set of init containers, 
while cluster administrators manage storage plugins, networking plugins, and container runtime handlers. 
Therefore, there is a need for an improved mechanism to provide cluster administrators with a clear and 
comprehensive view of pod sandbox creation completion and container readiness.

## What's the benefit?

1.Improved Visibility: Cluster administrators gain a clearer and more comprehensive view of pod sandbox creation completion and container readiness. 
This enhanced visibility allows them to make better-informed decisions and troubleshoot issues more effectively.

> 2.Metric Collection and Monitoring: Monitoring services can leverage the fields associated with the PodReadyToStartContainers condition to report sandbox creation state and latency. 

Reviewer 3: ```suggestion
2.Metric Collection and Monitoring: Monitoring services can leverage the fields associated with the PodReadyToStartContainers condition
to report sandbox creation state and latency. 
```

Author: updated.

> Metrics can be collected at per-pod cardinality or aggregated based on various properties of the pod, such as volumes, storage class of Persistent Volume Claims (PVCs), runtime class, 

Reviewer 3: ```suggestion
Metrics can be collected at per-pod cardinality or aggregated based on various
properties of the pod, such as volumes, storage class of Persistent Volume
Claims (PVCs), runtime class, 
```

Author: updated.

custom annotations for CNI and IPAM plugins, and arbitrary labels and annotations on pods. 
This enables comprehensive monitoring and analysis of pod readiness across the cluster.

> 3.Enhanced Troubleshooting: With a more accurate representation of pod sandbox creation and container readiness, cluster administrators can quickly identify and address any issues that may arise during the initialization process. 

Reviewer 3: ```suggestion
3.Enhanced Troubleshooting: With a more accurate representation of pod sandbox creation and container readiness,
cluster administrators can quickly identify and address any issues that may arise during the initialization process. 
```

Author: updated.

---
layout: blog
title: "Kubernetes 1.29: PodReadyToStartContainersCondition in beta"
date: 2023-11-13
slug: pod-ready-to-start-containers-condition-now-in-beta
---

**Author**: zefeng chen, Kevin Hannon(Red Hat)

With recent release of Kubernetes 1.29 the `PodReadyToStartContainers`[condition](/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions) is available by default.
The kubelet manages the value for that condition throughout a Pod's lifecycle, 
in the status field of a pod. The kubelet will use the PodReadyToStartContainers 
condition to accurately surface the initialization state of a pod,

> From the perspective of pod sandbox creation and network configuration by a container runtime

Reviewer 2: ```suggestion
from the perspective of pod sandbox creation and network configuration by a container runtime.
```

Author: updated

## What's the motivation and why we need it?

> Cluster administrators currently do not have a clear and easily accessible way to view the status of successfully scheduled pods 

Reviewer 2: I would probably drop the successfully scheduled pods and say something like:

Cluster administrators currently do not have a clear and easily accessible way to view the completion of pod sandbox creation.

Author: updated

## What's the benefit?

1.Improved Visibility: Cluster administrators gain a clearer and more comprehensive view of pod sandbox creation completion and container readiness. 
This enhanced visibility allows them to make better-informed decisions and troubleshoot issues more effectively.

2.Metric Collection and Monitoring: Monitoring services can leverage the fields associated with the PodReadyToStartContainers condition to 
report sandbox creation state and latency. 
Metrics can be collected at per-pod cardinality or aggregated based on various 
properties of the pod, such as volumes, storage class of Persistent Volume 
Claims (PVCs), runtime class, 
custom annotations for CNI and IPAM plugins, and arbitrary labels and annotations on pods. 
This enables comprehensive monitoring and analysis of pod readiness across the cluster.

3.Enhanced Troubleshooting: With a more accurate representation of pod sandbox creation and container readiness, 
cluster administrators can quickly identify and address any issues that may arise during the initialization process. 
This leads to improved troubleshooting capabilities and reduced downtime.

### What’s next?

> Depending on feedback and adoption, the Kubernetes team plans to promote

Reviewer 2: We promoted this feature to beta in 1.29. We want to get more users so we can promote to GA eventually.

Reviewer 2: Generally it will be on for all users so we want feedback on whether or not this feature is useful and other use cases for it.

Author: add little combine

the `PodReadyToStartContainersCondition` condition to Beta in 1.29.

### How can I learn more?

> Please check out the

Reviewer 2: Maybe link the KEP also?

Reviewer 2: For example, maybe a sentence like the following:
For more detailed overview of the feature, please read [KEP-3085](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md).

Also maybe mention other blog posts around this feature ie https://kubernetes.io/blog/2022/09/14/pod-has-network-condition/

But maybe that is too much..

[documentation](/docs/concepts/workloads/pods/pod-lifecycle/) for the
`PodReadyToStartContainersCondition` to learn more about it and how it fits in relation to
other pod conditions.

> ### How to get involved?

Author: this one is follow from the GRPC probe one, will check some other post to see what's difference
